### ISSUE TITLE
deadlock between `RmvTime` and timer thread

### ISSUE THREAD

**Issue by rakslice** [kanjitalk755/macemu#65]
While testing some other changes in SheepShaver on Linux, I ran into this situation:

SheepShaver is hung not responding to input or redrawing.

The main thread is in
```
(gdb) bt
#0  __lll_lock_wait (futex=futex@entry=0xaaaab1b477e0 <wakeup_time_lock>, private=0) at lowlevellock.c:52
#1  0x0000ffff823a4b98 in __GI___pthread_mutex_lock (mutex=mutex@entry=0xaaaab1b477e0 <wakeup_time_lock>) at pthread_mutex_lock.c:80
#2  0x0000aaaab1a93390 in RmvTime(unsigned int) (tm=271287612) at ../timer.cpp:379
#3  0x0000aaaab1a90674 in EmulOp(M68kRegisters*, uint32, int) (r=r@entry=0xffffc2531490, pc=<optimized out>, selector=<optimized out>) at ../emul_op.cpp:220
#4  0x0000aaaab1ad9e38 in sheepshaver_cpu::execute_emul_op(unsigned int) (this=0xaaaae9fef360, emul_op=<optimized out>) at ../kpx_cpu/src/cpu/ppc/ppc-cpu.hpp:110
#5  0x0000000020a12068 in  ()
#6  0x0000aaaab1accc30 in basic_dyngen::execute(unsigned char*) (this=0xaaaaea0efbb0, this=0xaaaaea0efbb0, entry_point=<optimized out>)
    at ../kpx_cpu/src/cpu/jit/basic-dyngen.hpp:293
#7  powerpc_cpu::execute(unsigned int) (this=0xaaaae9fef360, entry=540092988) at ../kpx_cpu/src/cpu/ppc/ppc-cpu.cpp:594
#8  0x0000aaaab1accc30 in basic_dyngen::execute(unsigned char*) (this=0xaaaab1be8c30, this=0xaaaab1be8c30, entry_point=<optimized out>)
    at ../kpx_cpu/src/cpu/jit/basic-dyngen.hpp:293
#9  powerpc_cpu::execute(unsigned int) (this=0xaaaab1ae83e0, entry=3929737344) at ../kpx_cpu/src/cpu/ppc/ppc-cpu.cpp:594
#10 0x0000aaaaea0efbb0 in  ()
```

```
#3  0x0000aaaab1a90674 in EmulOp (r=r@entry=0xffffc2531490, pc=<optimized out>, selector=<optimized out>) at ../emul_op.cpp:220
220                             r->d[0] = RmvTime(r->a[0]);
(gdb) list
215
216                     case OP_INSTIME:                        // InsTime() replacement
217                             r->d[0] = InsTime(r->a[0], r->d[1]);
218                             break;
219                     case OP_RMVTIME:                        // RmvTime() replacement
220                             r->d[0] = RmvTime(r->a[0]);
221                             break;
222                     case OP_PRIMETIME:                      // PrimeTime() replacement
223                             r->d[0] = PrimeTime(r->a[0], r->d[0]);
224                             break;
```

```
#2  0x0000aaaab1a93390 in RmvTime (tm=271287612) at ../timer.cpp:379
379             pthread_mutex_lock(&wakeup_time_lock);
(gdb) list
374             semaphore_wait(wakeup_time_sem);
375             thread_suspend(timer_thread);
376     #endif
377     #if PRECISE_TIMING_POSIX
378             timer_thread_suspend();
379             pthread_mutex_lock(&wakeup_time_lock);
380     #endif
381             if (ReadMacInt16(tm + qType) & 0x8000) {
382
383                     // Yes, make task inactive and remove it from the Time Manager queue
```

i.e. the main thread has run `RmvTime` which has just suspended the timer thread and is now waiting on `wakeup_time_lock`.

Meanwhile in the timer thread

```
(gdb) thread 16
[Switching to thread 16 (Thread 0xffff227fbe60 (LWP 15700))]
#0  0x0000ffff810f944c in __GI___sigsuspend (set=0xaaaab1b47848 <suspend_handler_mask>) at ../sysdeps/unix/sysv/linux/sigsuspend.c:26
26      ../sysdeps/unix/sysv/linux/sigsuspend.c: No such file or directory.
(gdb) bt
#0  0x0000ffff810f944c in __GI___sigsuspend (set=0xaaaab1b47848 <suspend_handler_mask>) at ../sysdeps/unix/sysv/linux/sigsuspend.c:26
#1  0x0000ffff824105b8 in <signal handler called> ()
#2  0x0000aaaab1a92f78 in timer_func(void*) (arg=<optimized out>) at ../timer.cpp:593
#3  0x0000ffff823a24fc in start_thread (arg=0xffffc2531a9f) at pthread_create.c:477
#4  0x0000ffff811960cc in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:78
(gdb) up
#1  <signal handler called>
(gdb) up
#2  0x0000aaaab1a92f78 in timer_func (arg=<optimized out>) at ../timer.cpp:593
593                             wakeup_time = wakeup_time_max;
(gdb) list
588                     timer_current_time(system_time);
589                     if (timer_cmp_time(wakeup_time, system_time) < 0) {
590
591                             // Timer expired, trigger interrupt
592                             pthread_mutex_lock(&wakeup_time_lock);
593                             wakeup_time = wakeup_time_max;
594                             pthread_mutex_unlock(&wakeup_time_lock);
595                             SetInterruptFlag(INTFLAG_TIMER);
596                             TriggerInterrupt();
597                     }
```

The timer thread is in the middle of the critical section on `wakeup_time_lock` but it has received the suspend from the main thread's `timer_thread_suspend()` and is now suspended so it is never going to release the lock.

A deadlock.

**Comment by rakslice** [comment]
Looking at the code in `timer.cpp`, the BeOS and Mach cases next to it already suspend the thread after the lock is acquired...
https://github.com/kanjitalk755/macemu/blob/3ce86cee1f62fd7a1179eec2cf0265872a2da4e4/SheepShaver/src/timer.cpp#L369

**Comment by rakslice** [comment]
It seems the order was switched around for POSIX as part of the change from semaphore to mutex in https://github.com/cebix/macemu/commit/dd7b09ebe872adb9fbf1a7e4352c780759321f41

I don't understand this remark in the commit message there:

> XXX: in the past, the sem_post in sigsuspend_handler would not release the
> calling thread thus entering into a mutual waiting condition.

**Comment by rakslice** [comment]
There are some changes related to the signals also in this commit. Perhaps the remark relates to that? I don't know if `wakeup_time_lock` is affected by anything around the signal; it is not directly referenced in `sigsuspend_handler`.

**Comment by kanjitalk755** [comment]
Tell me how to reproduce this issue.

**Comment by rakslice** [comment]
This happens to me in relatively short order (a couple minutes of use) when I use the bundled Netscape Communicator 4.8 in Mac OS 9.0.4 to browse an HTTP server on my local LAN on a dual core host system.

I'm not sure how useful that is, as I would expect it is very timing sensitive.  Perhaps I could make a small program to stress test with Time Manager `InsTime`/`RmvTime` cycles inside the emulated Mac?

**Comment by kanjitalk755** [comment]
I have found a case of deadlock with a low probability.
Test the following branch:
https://github.com/kanjitalk755/macemu/tree/test_timer_lock

**Comment by rakslice** [comment]
The test_timer_lock branch version tested ok.

Additionally I put together a program to exercise `InsTime`/`RmvTime`:

(source)
https://gist.github.com/rakslice/f2329427874e38edc649f4b24bd7328c
(build)
[time_mgr_stress_test.hfs.zip](https://github.com/kanjitalk755/macemu/files/5504392/time_mgr_stress_test.hfs.zip)

This hits the bug in relatively short order in the current version. It runs without issue on the test_timer_lock branch version.

Ironically, THINK C 5 itself doesn't work in SheepShaver but that's another story.

**Comment by kanjitalk755** [comment]
Merged to the master.
